contract/p/gnoswap/version_manager/version_manager.gno

```diff
@@ -0,0 +1,264 @@
+//   - Zero-downtime upgrades through hot-swapping
+//
+// Architecture Pattern: Strategy + Plugin Architecture
+package version_manager
+
+import (
+	"chain"
+	"chain/runtime"
+	"errors"
+	"strings"
+
+	"gno.land/p/gnoswap/store"
+	"gno.land/p/nt/avl"
+	"gno.land/p/nt/avl/rotree"
+	"gno.land/p/nt/ufmt"
+)
+
+// versionManager is the concrete implementation of VersionManager interface.
+// It manages multiple versioned implementations of a domain (e.g., protocol_fee/v1, protocol_fee/v2)
+// and controls their access permissions to a shared KVStore.
+//
+// The manager maintains a single active implementation with write access while keeping
+// previous versions available in read-only mode for backward compatibility.
+type versionManager struct {
+	// initializers stores registered initializer functions keyed by package path
+	// Each initializer bootstraps a specific version's implementation
+	initializers *avl.Tree
+
+	// domainKVStore is the shared storage layer accessible by all versions
+	// Access permissions are managed dynamically based on the active implementation
+	domainKVStore store.KVStore
+
+	// initializeDomainStoreFn wraps the KVStore into domain-specific storage interface
+	// This abstraction decouples the version manager from domain-specific storage implementations
+	initializeDomainStoreFn func(kvStore store.KVStore) any
+
+	// domainPath defines the base path for this domain (e.g., "gno.land/r/gnoswap/protocol_fee")
+	// Used for security validation to ensure only authorized packages can register
+	domainPath string
+
+	// currentPackagePath holds the package path of the active implementation
+	// (e.g., "gno.land/r/gnoswap/protocol_fee/v2")
+	currentPackagePath string
+
+	// currentImplementation is the active version's instance
+	// Only this implementation has write access to domainKVStore
+	currentImplementation any
+}
+
+// RegisterInitializer registers a new version implementation for the domain.
+// This method must be called by each version package (e.g., v1, v2) during initialization.
+//
+// The registration process:
+//  1. Validates the caller is within the authorized domain path
+//  2. Stores the initializer function for later version switching
+//  3. Handles first-time registration differently from subsequent ones:
+//     - First registration: Activates the version with write access
+//     - Subsequent registrations: Grants read-only access only
+//
+// Parameters:
+//   - initializer: A function that receives a storage interface and returns an implementation instance
+//
+// Returns:
+//   - error: If caller is unauthorized, already registered, or permission setup fails
+//
+// Security: Only packages under the domainPath prefix can register (enforced by isContainDomainPath)
+func (vm *versionManager) RegisterInitializer(initializer func(store any) any) error {
+	// Ensure the caller is within the domain path (e.g., protocol_fee/v1, protocol_fee/v2)
+	if !vm.isContainDomainPath() {
```

Comment:
Clever; by constraining it to the path it also ensures the deployment domain (ex. `gno.land/r/gnoswap`) is a protocol-level guard